src/GafferScene/IECoreScenePreview/MeshAlgoTessellate.cpp

```diff
@@ -0,0 +1,1491 @@
+		return 0;
+	}
+	else if( nVerts == 3 )
+	{
+		// One triangle in the center of the odd tessellation of a tri
+		return 1;
+	}
+	else
+	{
+		// For other odd tessellations, we get one center cap which is triangulated.
+		return nVerts;
+	}
+}
+
+} // namespace
+
+MeshPrimitivePtr MeshAlgo::tessellateMesh(
+	const MeshPrimitive &inputMesh, int divisions, const IECore::Canceller *canceller,
+	bool calculateNormals, bool tessellatePolygons,
+	SubdivisionScheme scheme
+)
+{
+	const int tessUniformRate = divisions + 1;
+
+	OpenSubdiv::Sdc::SchemeType osScheme = OpenSubdiv::Sdc::SCHEME_CATMARK;
+	switch( scheme )
+	{
+		case SubdivisionScheme::Bilinear:
+			osScheme = OpenSubdiv::Sdc::SCHEME_BILINEAR;
+			break;
+		case SubdivisionScheme::CatmullClark:
+			osScheme = OpenSubdiv::Sdc::SCHEME_CATMARK;
+			break;
+		case SubdivisionScheme::Loop:
+			osScheme = OpenSubdiv::Sdc::SCHEME_LOOP;
+			break;
+		default:
+			if( inputMesh.interpolation() == "linear" )
+			{
+				osScheme = OpenSubdiv::Sdc::SCHEME_BILINEAR;
+			}
+			else if( inputMesh.interpolation() == "catmullClark" )
+			{
+				osScheme = OpenSubdiv::Sdc::SCHEME_CATMARK;
+			}
+			else if( inputMesh.interpolation() == "loop" )
+			{
+				osScheme = OpenSubdiv::Sdc::SCHEME_LOOP;
+			}
+			else
+			{
+				throw Exception( "Unknown subdivision scheme on mesh: " + inputMesh.interpolation() );
+			}
+			break;
+	}
+
+	if( osScheme == OpenSubdiv::Sdc::SCHEME_LOOP && inputMesh.maxVerticesPerFace() > 3 )
+	{
+		throw Exception( "Loop subdivision can only be applied to triangle meshes ");
+	}
+
+	// Create PrimvarSetups for all the primvars we need to interpolate
+
+	if( !inputMesh.variableData<V3fVectorData>( "P", PrimitiveVariable::Vertex ) )
+	{
+		throw IECore::Exception( "Mesh must have V3f P primvar." );
+	}
+	if( !inputMesh.isPrimitiveVariableValid( inputMesh.variables.at( "P" ) ) )
+	{
+		throw IECore::Exception( "P primvar is invalid." );
+	}
+	PrimvarSetup posPrimvarSetup( "P", inputMesh.variables.at( "P" ) );
+
+	std::vector< PrimvarSetup > vertexPrimvarSetups;
+	std::vector< PrimvarSetup > uniformPrimvarSetups;
+	std::vector< PrimvarSetup > faceVaryingPrimvarSetups;
+	setupVariables(
+		inputMesh, calculateNormals,
+		posPrimvarSetup, vertexPrimvarSetups, uniformPrimvarSetups, faceVaryingPrimvarSetups, canceller
+	);
+
+
+	// These subdiv options hold all the tricky boundary settings
+	OpenSubdiv::Sdc::Options options;
+	options.SetVtxBoundaryInterpolation(OpenSubdiv::Sdc::Options::VTX_BOUNDARY_EDGE_AND_CORNER);
+
+	// Choosing a reasonable default here is actually tricky - the options are pretty confusing, and
+	// don't match between different packages ( and it seems like often artists may not actually be
+	// getting exactly what they expect ). This was chosen to match Arnold's default. Interestingly,
+	// it's also 30% faster than FVAR_LINEAR_BOUNDARIES.
+	options.SetFVarLinearInterpolation(OpenSubdiv::Sdc::Options::FVAR_LINEAR_CORNERS_ONLY);
```

Comment:
This whole thing feels like an absolute mess, but I guess potentially adding controls for this will immediately be the next thing after this PR is done ... which doesn't really clarify anything, but hopefully makes it the user's problem to get this wrong and confuse themselves.